File: visualization-frontend/src/model/node.js
import {assert} from './util';

export class Node {

	constructor(number, clause, inferenceRule, parents, statistics, isFromPreprocessing) {
		this.checkAssertions(number, clause, inferenceRule, parents, statistics, isFromPreprocessing);

		this.number = number;
        this.clause = clause; // TODO: remove quotes
        this.inferenceRule = inferenceRule;
        this.parents = parents;
        this.statistics = statistics;
        this.isFromPreprocessing = isFromPreprocessing;
        this.newTime = undefined;
        this.passiveTime = undefined;
		this.activeTime = undefined;
	}

	setNewTime(time) {
		assert(typeof time === "number");
		assert(this.newTime === undefined);
		assert(this.passiveTime === undefined);
		assert(this.activeTime === undefined);
		this.newTime = time;
	}

	setPassiveTime(time) {
		assert(typeof time === "number");
		assert(this.newTime !== undefined);
		assert(this.passiveTime === undefined);
		assert(this.activeTime === undefined);
		this.passiveTime = time;
	}

	setActiveTime(time) {
		assert(typeof time === "number");
		assert(this.newTime !== undefined);
		assert(this.passiveTime !== undefined);
		assert(this.activeTime === undefined);
		this.activeTime = time;
	}

	checkAssertions(number, clause, inferenceRule, parents, statistics, isFromPreprocessing) {
		// cf. discussion https://stackoverflow.com/questions/899574/what-is-the-difference-between-typeof-and-instanceof-and-when-should-one-be-used\  
		assert(typeof number === "number");
        assert(typeof clause === "string");
		assert(typeof inferenceRule === "string");
		assert(typeof isFromPreprocessing === "boolean");
		assert(Array.isArray(parents));
		assert(Array.isArray(statistics));
		parents.forEach(parent => {
			assert(typeof parent === "number");
		});
		statistics.forEach(stat => {
			assert(typeof stat === "number");
		});
	}

	toString() {
		return this.clause;
	}
}

File: visualization-backend/proof_visualization/model/node.py
"""Data object representing a line of proof output"""

import proof_visualization.model.util as util


class Node:
    def __init__(self, number, clause, inference_rule, parents, statistics, is_from_preprocessing):
        self._check_assertions(number, clause, inference_rule, parents, statistics, is_from_preprocessing)

        self.number = number
        self.clause = util.remove_quotes(clause)
        self.inference_rule = inference_rule
        self.parents = parents

        self.statistics = statistics

        self.is_from_preprocessing = is_from_preprocessing
        self.new_time = None
        self.passive_time = None
        self.active_time = None

    def set_new_time(self, line_number):
        assert isinstance(line_number, int)
        assert self.new_time is None
        assert self.passive_time is None
        assert self.active_time is None
        self.new_time = line_number

    def set_passive_time(self, line_number):
        assert isinstance(line_number, int)
        assert self.new_time is not None
        assert self.passive_time is None
        assert self.active_time is None
        self.passive_time = line_number

    def set_active_time(self, line_number):
        assert isinstance(line_number, int)
        assert self.new_time is not None
        assert self.passive_time is not None
        assert self.active_time is None
        self.active_time = line_number

    def __str__(self):
        return self.clause

    def __repr__(self):
        return self.clause

    @staticmethod
    def _check_assertions(number, clause, inference_rule, parents, statistics, is_from_preprocessing):
        assert isinstance(number, int)
        assert isinstance(clause, str)
        assert isinstance(inference_rule, str)
        assert isinstance(parents, list)
        assert isinstance(statistics, list)
        for parent in parents:
            assert isinstance(parent, int)
        for stat in statistics:
            assert isinstance(stat, int)
        assert isinstance(is_from_preprocessing, bool)
